crates/lang_srv/src/server.rs

```diff
@@ -187,22 +296,264 @@ impl LanguageServer for RocLs {
+        match resp {
+            CompletionResponse::Array(list) => list.into_iter(),
+            CompletionResponse::List(list) => list.items.into_iter(),
+        }
+        .map(|item| item.label)
+        .collect::<Vec<_>>()
+    }
+    ///Gets completion and returns only the label for each completion
+    async fn get_completion_labels(
+        reg: &Registry,
+        url: &Url,
+        position: Position,
+    ) -> Option<Vec<String>> {
+        reg.completion_items(url, position)
+            .await
+            .map(completion_resp_to_labels)
+    }
+
+    const DOC_LIT: &str = r#"
+app "fizz-buzz"
+    packages { pf: "https://github.com/roc-lang/basic-cli/releases/download/0.5.0/Cufzl36_SnJ4QbOoEmiJ5dIpUxBvdB3NEySvuH82Wio.tar.br" }
+    imports [pf.Stdout,pf.Task.{ Task, await },]
+    provides [main] to pf
+"#;
+    static INIT: Once = Once::new();
+    async fn test_setup(doc: String) -> (RocServerState, Url) {
+        INIT.call_once(|| {
+            env_logger::builder()
+                .is_test(true)
+                .filter_level(log::LevelFilter::Trace)
+                .init();
+        });
+        let url = Url::parse("file:/test.roc").unwrap();
+
+        let inner = RocServerState::new();
+        //setup the file
+        inner.change(&url, doc, 0).await.unwrap();
+        (inner, url)
+    }
+
+    #[tokio::test]
+    async fn test_completion_with_changes() {
+        let doc = DOC_LIT.to_string()
+            + r#"rec=\a,b->{one:{potato:\d->d,leak:59},two:b}
+rectest= 
+  value= rec 1 2
+  va"#;
+        let (inner, url) = test_setup(doc.clone()).await;
+        static INNER_CELL: OnceLock<RocServerState> = OnceLock::new();
+        INNER_CELL.set(inner).unwrap();
+        static URL_CELL: OnceLock<Url> = OnceLock::new();
+        URL_CELL.set(url).unwrap();
+
+        let inner = INNER_CELL.get().unwrap();
+        let url = URL_CELL.get().unwrap();
+        let position = Position::new(8, 8);
+        //setup the file
+        inner.change(&url, doc.clone(), 1).await.unwrap();
+
+        //apply a sequence of changes back to back
+        let a1 = spawn(inner.change(&url, doc.clone() + "l", 2));
+        let a2 = spawn(inner.change(&url, doc.clone() + "lu", 3));
+        let a3 = spawn(inner.change(&url, doc.clone() + "lue", 4));
+        let a4 = spawn(inner.change(&url, doc.clone() + "lue.", 5));
+        //start a completion that would only work if all changes have been applied
+        let comp = spawn(async move {
+            let reg = inner.registry().await;
+            get_completion_labels(reg, &url, position).await
+        });
+        // Simulate two changes coming in with a slight delay
+        let a = spawn(inner.change(&url, doc.clone() + "lue.o", 6));
+        tokio::time::sleep(Duration::from_millis(100)).await;
+        let rest = spawn(inner.change(&url, doc.clone() + "lue.on", 7));
+
+        let done = join!(a1, a2, a3, a4, comp, a, rest);
+
+        expect![[r#"
+            (
```

Comment:
I would drop the `Ok(Err(_))` variants here. I don't think they're super useful, and I suspect they will not be deterministic. On a similar note, let's get rid of the synthetic sleep in the test for now.

Reply:
I would like to ensure that we have tests that validate that
1: Sequential changes are skipped without analysis
2: a change arriving before analysis is done will cause the previous analysis to be discarded

On an extremely fast computer it's conceivable the analysis could finish before 100ms 

I can implement an env var that sets the debounce delay for starting analysis, and during this test i could set it to 0ms and only delay for 10ms between the messages, the analysis will never finish that fast.

Is that a good enough solution?

Reply:
I would rather refactor the implementation to dependency-inject a function to analyze the module, and in the tests make that function block until a later analysis requests is issued. I want to avoid sleeps entirely because it’s very likely to start flaking at some point (even if it doesn’t today), and the only way to fix it at that point will be to change the testing strategy anyway.

Reply:
That's fair, ill take the test out for now and do another PR adding it back in soon